Minting edge cases and additional testing

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->

Edge cases addressed:

- [x] We now reject asset names that are too long.
- [x] We now reject asset quantities that are out of bounds.
- [x] We now accept empty asset names.
- [x] We now handle minting to foreign addresses.

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number
ADP-1542
<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
asset name too long

paweljakubas committed:
wrong token quantity

paweljakubas committed:
introduce optional token name with test

paweljakubas committed:
add preliminary integration test for minting into foreign address

paweljakubas committed:
introduce optional address when minting and add first version of minting out to selection

paweljakubas committed:
update integration test

paweljakubas committed:
add token map coalescence functionality within the same address

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -3327,6 +3315,126 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
+                --, expectField
+                --        (#balance . #available . #getQuantity)
+                --        (`shouldBe` minutxo)
+                -- expected: 999978 (minutxo)
```
this one I don't understand at this moment @jonathanknowles

Anviking replied:
It's probably related to the minUTxO value increasing when the output contains assets. It's 344820 lovelace greater, which divided by 34482 `lovelacePerUTxOWord` is 10 words or 80 bytes, which sounds like it could be the size of the assets in the output.

paweljakubas replied:
thanks Johannes! @Anviking  indeed, minimum utxo needs to be adjusted here. Now, everything is good!

paweljakubas committed:
hlint and stylish

paweljakubas committed:
add fix for minUTxO value due to TxOuts with assets

jonathanknowles committed:
Format for consistency with team coding standards.

- use 4-character indents.
- avoid variable-length indentation.
- limit line lengths to 80 characters where practical.
- use only a single blank line between top-level definitions.

https://input-output-hk.github.io/adrestia/code/Coding-Standards

jonathanknowles committed:
Rename `CreatedTransactionWithTooLongAssetName``.

Rename `CreatedTransactionWithTooLongAssetName` to `AssetNameTooLong`.

This error message implies that we've actually created a transaction.

But if an asset name is too long, then it isn't possble for us to have
created a transaction. Therefore, we can't say that we've created one.

Additionally, if an asset name is too long, then it's too long in all
contexts.

Therefore, it's enough to say that "the asset name is too long": we
don't actually need to state the context.

We already use this style in other areas. For example:

- AssetNotPresent
- WalletAlreadyExists

jonathanknowles committed:
Sort `ApiErrorCode` constructors into alphabetical order.

It's easier to identify and reuse an existing constructor if the list is
sorted into alphabetical order.

jonathanknowles committed:
Rename `CreatedTransactionWithIncorrectAssetQuantity`.

Rename `CreatedTransactionWithIncorrectAssetQuantity` to
`AssetQuantityOutOfBounds`.

If an asset quantity is out of bounds, then it's impossible for us to
say we have created a transaction.

Therefore, it's enough to say "asset quantity out of bounds".

The `AssetQuantityOutOfBounds` error name is consistent with other
existing error names:

- `AssetNameTooLong`
- `AssetNotPresent`
- `AssetQuantityOutOfBounds`

jonathanknowles committed:
Revise error string for `ErrConstructTxAssetNameTooLong`.

The upper limit is 32 bytes, but this does not necessarily correspond to
16 characters.

According to the UTF-8 standard, Unicode code points can require between
1 and 4 bytes to encode.

jonathanknowles committed:
Revise error string for `ErrConstructTxAssetQuantityOutOfBounds`.

Mention that the upper limit is equal to 2^63-1.

jonathanknowles committed:
Revise description of `receivingAddress` in `ApiMintData`.

Emphasise that when an address is not specified, that minted assets will
be returned to the wallet as change, and that change addresses will
assigned automatically.

jonathanknowles committed:
Revise definition of `maxTokenQuantity`.

- Move this constant to `Primitive.Types.Tx`, so that it is located
  together with similar constants.

- Redefine it in terms of `Int64` rather that `Int`, which may be
  platform specific.

- Rename it to `txMintBurnMaxTokenQuantity`, to emphasise that it only
  relates to minting and burning.

jonathanknowles committed:
Rename `AssetQuantityOutOfBounds`.

Rename `AssetQuantityOutOfBounds` to `MintOrBurnAssetQuantityOutOfBounds`.

The upper limit for a token quantity in the `mint` field of a
transaction is actually different from the upper limit for a token
quantity in an ordinary transaction output.

So we use a more specific name to emphasise this difference.

jonathanknowles committed:
Rename `maxLengthTokenName` to `tokenNameMaxLength`.

This name was a bit confusing, because it could be interpreted as being
"a token name of the maximum length", when in actuality it is "the
maximum length of a token name".

jonathanknowles committed:
Change type of `txMintBurnMaxTokenQuantity` to `TokenQuantity`.

This is consistant with other constants in the `Tx` module.

paweljakubas committed:
treat policyXPub in more conservative way
